internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
+			s.transition(stateBackoffWait)
+			time.AfterFunc(duration, func() { logError(s.backoffWaitElapsed()) })
+
+		default:
+			return fmt.Errorf("internal error: unexpected action %q", action)
 		}
-		return fmt.Errorf("cannot start service: exited quickly with code %d", cmd.ProcessState.ExitCode())
+
+	case stateTerminating, stateKilling:
+		s.stopped <- nil
+		s.transition(stateStopped)
+
+	default:
+		return fmt.Errorf("exited invalid in state %q", s.state)
 	}
-	// unreachable
+	return nil
 }
 
-func (m *ServiceManager) doStop(task *state.Task, tomb *tomb.Tomb) error {
-	releasePlan, err := m.acquirePlan()
-	if err != nil {
-		return err
+// exitCode returns the exit code of the given command, or 128+signal if it
+// exited via a signal.
+func exitCode(cmd *exec.Cmd) int {
+	status, ok := cmd.ProcessState.Sys().(syscall.WaitStatus)
+	if ok {
+		if status.Signaled() {
+			return 128 + int(status.Signal())
+		}
+		return status.ExitStatus()
 	}
-	defer releasePlan()
+	return cmd.ProcessState.ExitCode()
+}
 
-	m.state.Lock()
-	req, err := TaskServiceRequest(task)
-	m.state.Unlock()
-	if err != nil {
-		return err
+// getAction returns the correct action to perform from the plan and whether
+// or not the service exited with a success exit code (0).
+func getAction(config *plan.Service, success bool) (action plan.ServiceAction, onType string) {
+	switch {
+	case !success && config.OnFailure != "":
+		return config.OnFailure, "on-failure"
+	case success && config.OnSuccess != "":
+		return config.OnSuccess, "on-success"
+	default:
+		onExit := config.OnExit
+		if onExit == "" {
+			onExit = plan.ActionRestart // default for "on-exit"
+		}
+		return onExit, "on-exit"
 	}
+}
 
-	active, ok := m.services[req.Name]
-	if !ok {
-		// Already stopped
-		m.state.Lock()
-		task.Logf("Service %q already stopped.", req.Name)
-		m.state.Unlock()
-		return nil
+// sendSignal sends the given signal to a running service.
+func (s *service) sendSignal(signal string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateStarting, stateRunning:
+		err := syscall.Kill(-s.cmd.Process.Pid, unix.SignalNum(signal))
+		if err != nil {
+			return err
+		}
+
+	case stateBackoffWait, stateTerminating, stateKilling, stateStopped:
+		return fmt.Errorf("cannot send signal while service is stopped or stopping")
+
+	default:
+		return fmt.Errorf("sendSignal invalid in state %q", s.state)
 	}
-	cmd := active.cmd
+	return nil
+}
 
-	releasePlan()
+// stop is called to stop a running (and backing off) service.
+func (s *service) stop() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
 
-	syscall.Kill(-cmd.Process.Pid, syscall.SIGTERM)
-	//cmd.Process.Signal(syscall.SIGTERM)
+	switch s.state {
+	case stateRunning:
+		// First send SIGTERM to try to terminate it gracefully.
+		err := syscall.Kill(-s.cmd.Process.Pid, syscall.SIGTERM)
+		if err != nil {
+			logger.Noticef("cannot send SIGTERM to process: %v", err)
+		}
+		s.stopped = make(chan error, 1)
```

Comment:
I haven't looked into it, but given the complexity involved here, we need to make sure that this buffering is enough, and there's no risk of attempting to send twice on this channel. Doing so would mean a blocked goroutine and a frozen system.

Reply:
Yes, good point. This channel is sent to in state "terminating" or "killing" (after SIGTERM or SIGKILL has been sent), or in state "killing" after the kill wait has elapsed. However, if the kill wait elapses it sends a stop error on the channel (but remains in "killing" state for lack of something better to do), so if the process exits after that (unlikely but possible) it'll mean a blocked goroutine. So a size of 2 is sufficient -- I've updated it.